workspace/notebook/py_sb_harmonised_nsip_invoice.json

```diff
@@ -0,0 +1,641 @@
+					"                F.coalesce(F.col(\"refundCreditNoteNumber\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"refundAmount\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"refundIssueDate\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"Migrated\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"ODTSourceSystem\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"SourceSystemID\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"IngestionDate\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"ValidTo\").cast(\"string\"), F.lit(\".\")),\n",
+					"                F.coalesce(F.col(\"NewIsActive\").cast(\"string\"), F.lit(\".\"))\n",
+					"            )\n",
+					"        )\n",
+					"    )\n",
+					"\n",
+					"    # Keep only latest record per group for MERGE (row_num == 1)\n",
+					"    # display(df)\n",
+					"    # latest_df = df.filter(F.col(\"row_num\") == 1)\n",
+					"\n",
+					"    # display(latest_df)\n",
+					"\n",
+					"    # Step 3: Merge updates back into Delta table\n",
+					"    target_table.alias(\"t\").merge(\n",
+					"        df.alias(\"s\"),\n",
+					"        \"t.caseId = s.caseId AND t.invoiceNumber = s.invoiceNumber AND t.IngestionDate = s.IngestionDate And t.NSIPProjectInfoInternalID=s.NSIPProjectInfoInternalID \" \n",
+					"    ).whenMatchedUpdate(\n",
+					"        set={\n",
+					"            \"IsActive\": \"s.NewIsActive\",\n",
+					"            \"ValidTo\": F.expr(\"CASE WHEN s.NewIsActive = 'N' THEN current_timestamp() ELSE t.ValidTo END\"),\n",
+					"            \"Migrated\": \"s.Migrated\",\n",
+					"            \"RowID\": \"s.RowID\"\n",
+					"        }\n",
+					"    ).execute()\n",
+					"    end_exec_time = datetime.now()\n",
+					"\n",
+					"    print(\"MERGE completed successfully. Latest record IsActive='Y', older ones='N'.\")\n",
+					"\n",
+					"except Exception as e:\n",
+					"    error_message = f\"Error during MERGE operation: {str(e)[:800]}\"\n",
+					"    end_exec_time = datetime.now()\n",
+					"    print(error_message)\n",
+					""
+				],
+				"execution_count": null
+			},
+			{
+				"cell_type": "code",
+				"source": [
+					"duration_seconds = (end_exec_time - start_exec_time).total_seconds()\n",
+					"activity_type = f\"{mssparkutils.runtime.context['currentNotebookName']} Notebook\"\n",
+					"stage = \"Success\" if not error_message else \"Failed\"\n",
+					"status_message = (\n",
+					"    f\"Successfully loaded data into {spark_table_final} table\"\n",
+					"    if not error_message\n",
+					"    else f\"Failed to load data from {spark_table_final} table\"\n",
+					")\n",
+					"status_code = \"200\" if stage == \"Success\" else \"500\"\n",
+					" \n",
+					"log_telemetry_and_exit(\n",
```

Comment:
Here you call log_telemetry but you never in the notebook have a cell to %run utils/py_applicationinsights which you have in the curated notebook but not here in Harmonised this will give a NameError for log_telemetry. Just add a cell after the imports %run utils/py_applicationinsights